### PR TITLE
Rgb shift

### LINKED ISSUE #941: [Feat] Kornia Augmentations tracker
## More Than Augmentation
In a longer vision, our augmentation module will be more focused on differentiability, with an aim of assisting both training and deployment. We can see there are more researches and applications in this area, with or without Kornia. This module will be the lead. In our next version, we gonna heavily improve:
- the API design and better readability.
- Advanced features like JIT and AMP.
- Better differentiability and more differentiable augmentations.

## 🚀 Feature
Specs for the `kornia.augmentation` module.

- Make it easy to contrib through `AugmentationBase2D`.
- Give support to `torch.jit.script` and `torch.float16`.
- Promote to embed the full code in the main class to keep readability.
- Documentation improvement:
  - Better looking
  - More examples for copying and pasting.
  - Provide a benchmark to notify users that we are better.
- Add `register_buffer` to save static parameters inside ```nn.Module```.

## Motivation

This should be the template to implement a custom augmentation

```python
class RandomHorizontalFlip(AugmentationBase2D):

    def compute_transformation(self, input: torch.Tensor) -> torch.Tensor:
        w: int = input.shape[-1]
        flip_mat: torch.Tensor = torch.tensor([[-1, 0, w - 1],
                                               [0, 1, 0],
                                               [0, 0, 1]], device=input.device, dtype=input.dtype)
        return flip_mat.repeat(input.size(0), 1, 1)

    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
        return kornia.hflip(input)
```

## User Stories
- _As a user, I want …, so that …_ 
- As a user, I want to decide whether the transforms return the transformation, so that I can use it to chain and potentially apply in sequential.
- As a user, I want to use the augmentations with half precision, so that I can also combine with AMP.
- As a user, I want to be able to run the augmentations with xla devices, so that I can run on TPUs.
- As a user, I want to be able to torchscript the augmentations, so that I can easily serialise and export.
- As a developer, I want to have easy examples of custom augmentations, so that I can contribute with more.

## TODO:
- [x] Refactor engine to deprecate `return_transform` to return only `torch.Tensor` @shijianjian 
  - Transforms will be accessible through class properties
- [ ] Enable torchscript for augmentations https://github.com/kornia/kornia/issues/477
- [x] Remove functional code @shijianjian 
- [x] Fix `RandResizedCrop` critical bug https://github.com/kornia/kornia/issues/927

## List of existing 2d transforms:
- [x] CenterCrop
- [x] ColorJitter
- [x] GaussianBlur
- [x] RandomAffine
- [x] RandomCrop
- [ ] RandomErasing
  - Gradients error. They computed on the non-cropped area than on the mask. Check the gradient image in the bottom. will be fixed in https://github.com/kornia/kornia/pull/873 
- [x] RandomGrayscale
- [x] RandomHorizontalFlip
- [x] RandomVerticalFlip
- [x] RandomMotionBlur
- [x] RandomPerspective
- [x] RandomResizedCrop
  - Broken: https://github.com/kornia/kornia/issues/927
- [x] RandomRotation
- [x] RandomSolarize
- [ ] RandomPosterize
  - Not differentiable (due to the basis of int8 operations). Grad_fn not implemented.
- [x] RandomSharpness
- [ ] RandomEqualize
  - Not backward-propagatable
  - Error message: “derivative for floor_divide is not implemented”
- [x] RandomMixUp
- [x] RandomCutMix

## List of wanted 2d transforms:
- **Adapt from Tormentor:** needs tormentor’s diamond square
  - [x] RandomelasticTransform  @edgarriba
  - [x] RandomThinPlateSpline  @edgarriba
  - [x] RandomFishEye  @edgarriba
    - Adopt from tormentor: https://github.com/anguelos/tormentor#create-custom-augmentation-class
  - [x] RandomPlasmaBrightness
  - [x] RandomPlasmaContrast
  - [ ] RandomPlasmaRgbBrightness
  - [x] RandomPlasmaShadow
  - [ ] ImageCorruption augmentations https://github.com/kornia/kornia/issues/1143
- **Adapt from Albumentations**
  - [x] RandomInvert https://github.com/kornia/kornia/pull/960 @edgarriba 
  - [x] RandomGaussianNoise https://github.com/kornia/kornia/pull/960 @edgarriba 
  - [x] RandomChannelShuffle https://github.com/kornia/kornia/pull/960 @edgarriba 
  - [ ] RandomGlassNoise
  - [x] RandomRGBShift
  - [ ] RandomClahe
  - [x] RandomBoxBlur
  - [ ] RandomMedianBlur
  - [ ] RandomChannelDropout
  - [ ] RandomDownScale
  - [ ] RandomRain -> need `draw_line`
  - [ ] RandomShadow -> need `draw_line`
  - [ ] RandomSnow https://github.com/UjjwalSaxena/Automold--Road-Augmentation-Library/blob/master/Automold.py#L330
  - [ ] RandomSunFlare -> need `draw_line`
- **Containers**
  - [x] VideoSequential
  - [x] OneOf ? or RandomChoice
  - [x] ToTensor https://github.com/kornia/kornia/issues/705
  - [x] Lambda
  - [x] RandomApply
  - [ ] AutoAugment ? (or similar) @shijianjian 
  - [ ] TTA @shijianjian 
  - [ ] Augmentation presets https://github.com/kornia/kornia/issues/1151

Link to discussion docs: https://docs.google.com/document/d/1DcWnoB5L-Ws0asOq9ntXHrCe0ny2NM2WCtLIoU50wQw/edit?usp=sharing

## Advanced features
 - [x] `AugmentationSequential` to support multi-data augmentations
 - [ ] Make `AugmentationSequential` end to end. 
     - [x] Bounding Box #1497 
     - [ ] Mask
     - [x] Keypoints
     - [ ] Label
- [x] Deprecate `return_transform` in favour of `inverse`.
- [ ] AutoAugment pipelines.

-----

## Please, anyone interested: comment below and volunteer yourself :smile:

### PR DESCRIPTION
**Pull request by Oleksandra2020**
#### Changes
Added RandomRGBShift

Fixes #941 

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Review by edgarriba — COMMENTED** [review]
Add some tests too

File: kornia/augmentation/_2d/intensity/random_rgb_shift.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                [[0.3850, 0.5623, 0.2636, 0.1328, 0.4005],
+                  [0.0000, 0.1584, 0.0000, 0.0000, 0.0000],
+                  [0.2914, 0.2663, 0.0000, 0.2163, 0.3397],
+                  [0.0021, 0.0843, 0.1811, 0.3754, 0.5453],
+                  [0.0000, 0.0000, 0.2617, 0.4795, 0.5003]]]])
+    """
+
+    def __init__(
+        self,
+        r_shift_limit: int = 0.5,
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_rgb_shift.py:81]:
```suggestion
        r_shift_limit: float = 0.5,
```

File: kornia/enhance/rgb_shift.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import torch
+
+
+def shift_rgb(image, r_shift, g_shift, b_shift):
+    """
+    Shift each image's channel by either r_shift for red, g_shift for green and b_shift for blue channels.
+    """
+
+    if not isinstance(image, torch.Tensor):
```
Comment (edgarriba) [inline kornia/enhance/rgb_shift.py:9]: I suggest To use the api below
https://github.com/kornia/kornia/blob/master/kornia/testing/__init__.py#L208

Also add the params and shapes the docstring

File: test/enhance/test_rgb_shift.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        image = utils.tensor_to_gradcheck_var(image)  # to var
+        assert gradcheck(kornia.enhance.shift_rgb, (image, r_shift, g_shift, b_shift), raise_exception=True)
+
+    def test_rgb_shift(self, device, dtype):
+        r_shift, g_shift, b_shift = 0.1, 0.2, -0.3
+        image = torch.tensor([[[[0.2, 0.]], [[0.3, 0.5]], [[0.4, 0.7]]]], device=device, dtype=dtype)
+        shifted = kornia.enhance.shift_rgb(image, r_shift, g_shift, b_shift)
+        expected = torch.tensor([[[[0.3, 0.1]], [[0.5, 0.7]], [[0.1, 0.4]]]], device=device, dtype=dtype)
+
+        return (shifted == expected).all()
```
Comment (edgarriba) [inline test/enhance/test_rgb_shift.py:50]:
```suggestion
        assert_close(shifted, expected)
```
Same for the others

File: kornia/augmentation/_2d/intensity/random_rgb_shift.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        g_shift_limit: int = 0.5,
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_rgb_shift.py:82]:
```suggestion
        g_shift_limit: float = 0.5,
```

File: kornia/augmentation/_2d/intensity/random_rgb_shift.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        b_shift_limit: int = 0.5,
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_rgb_shift.py:83]:
```suggestion
        b_shift_limit: float = 0.5,
```

**Review by edgarriba — CHANGES_REQUESTED** [review]
@shijianjian please review

File: kornia/enhance/rgb_shift.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import torch
+
+
+def shift_rgb(image, r_shift, g_shift, b_shift):
+    """
+    Shift each image's channel by either r_shift for red, g_shift for green and b_shift for blue channels.
+    """
```
Comment (shijianjian) [inline kornia/enhance/rgb_shift.py:7]:
```suggestion
def shift_rgb(image, r_shift, g_shift, b_shift):
    """Shift rgb channels.

    Shift each image's channel by either r_shift for red, g_shift for green and b_shift for blue channels.
    """
```

Typing as well.

**Review by shijianjian — COMMENTED** [review]
@Oleksandra2020 please fix the tests as well.

File: kornia/enhance/rgb_shift.py (hunk 6 above)
Comment (Oleksandra2020) [reply kornia/enhance/rgb_shift.py]: Thank you, changed.

File: kornia/augmentation/_2d/intensity/random_rgb_shift.py (hunk 5 above)
Comment (Oleksandra2020) [reply kornia/augmentation/_2d/intensity/random_rgb_shift.py]: Thank you, changed.

File: test/enhance/test_rgb_shift.py [hunk 7]
```diff
@@ -47,4 +50,9 @@ def test_rgb_shift(self, device, dtype):
         shifted = kornia.enhance.shift_rgb(image, r_shift, g_shift, b_shift)
         expected = torch.tensor([[[[0.3, 0.1]], [[0.5, 0.7]], [[0.1, 0.4]]]], device=device, dtype=dtype)
 
-        return (shifted == expected).all()
+        utils.assert_close(shifted, expected)
+
+
+if __name__ == "__main__":
```
Comment (edgarriba) [inline test/enhance/test_rgb_shift.py]: this should be removed: to run this particular test you can use `pytest -k TestRGBShift`

**Review by edgarriba — APPROVED** [review]

File: test/enhance/test_rgb_shift.py (hunk 7 above)
Comment (Oleksandra2020) [reply test/enhance/test_rgb_shift.py]: Yes, I've removed it, thanks.

**Comment by edgarriba** [comment]
@shijianjian i cannot see the conflicts with main. Any ideas ?

**Comment by edgarriba** [comment]
@Oleksandra2020 could you try to rebase with master ? i tried but for some reason i couldn't push the fixes

**Comment by edgarriba** [comment]
@also don't forget to add the functions to the docs https://github.com/kornia/kornia/blob/master/docs/source/augmentation.module.rst

**Review by shijianjian — COMMENTED** [review]
Please rebase from master. It is complaining but not showing the GUI.

File: kornia/enhance/rgb_shift.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import torch
+
+from kornia.testing import KORNIA_CHECK_IS_COLOR, KORNIA_CHECK_IS_TENSOR
+
+
+def shift_rgb(image, r_shift, g_shift, b_shift):
```
[the remark this answers is not in the document]
Comment (Oleksandra2020) [reply kornia/enhance/rgb_shift.py:6]: I did a rebase and it fetched the previous commits from the master.

**Comment by edgarriba** [comment]
@Oleksandra2020 something went wrong with the rebase. I suggest to revert until the last good commi and then `git merge master` (and solve the minor conflicts) and push again

File: kornia/enhance/shift_rgb.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    """
+
+    KORNIA_CHECK_IS_TENSOR(image)
+    KORNIA_CHECK_IS_COLOR(image, f"with shape {image.shape}")
+
+    shifts = [r_shift, g_shift, b_shift]
+
+    shifted = (image + torch.Tensor(shifts).view(1, 3, 1, 1).to(image)).clamp_(min=0, max=1)
+
+    return shifted
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/enhance/shift_rgb.py]: So, what the types should be for those parameters?

```python
shift_rgb(image, r_shift, g_shift, b_shift):
```

[the remark this answers is not in the document]
Comment (Oleksandra2020) [reply kornia/enhance/shift_rgb.py]: They are all torch.Tensor, I changed that

[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/enhance/shift_rgb.py]: Can `shifted = (image + torch.Tensor(shifts).view(1, 3, 1, 1).to(image)).clamp_(min=0, max=1)` be written in `torch` operations? Creating a new tensor here might lead to ONNX errors since it will loose track on the gradients from those 3 source tensors.

[the remark this answers is not in the document]
Comment (Oleksandra2020) [reply kornia/enhance/shift_rgb.py]: It can be done like this then:

```
shifted = torch.empty_like(image)
for i, shift in enumerate(shifts):
    shifted[:, i, :, :] = (image[:, i, :, :] + shift).clamp_(min=0, max=1)
```

[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/enhance/shift_rgb.py]: Still a new tensor created. To avoid that, would `shifted = (image + torch.stack(shifts).view(1, 3, 1, 1).to(image)).clamp_(min=0, max=1)` work?

[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/enhance/shift_rgb.py]: The best way to know it is adding onnx and torchscript test

[the remark this answers is not in the document]
Comment (Oleksandra2020) [reply kornia/enhance/shift_rgb.py]: It works with torch.stack

[the remark this answers is not in the document]
Comment (Oleksandra2020) [reply kornia/enhance/shift_rgb.py]: Yes, but what do you mean by onnx conversion? Does this test have to create a PyTorch model and convert it to ONNX and then check if there will be any onnx errors because of RGB shifting?
